curve25519-dalek/src/backend/serial/risc0/scalar.rs

```diff
@@ -0,0 +1,263 @@
+//! with RISC0 Acceleration
+
+use core::fmt::Debug;
+use crypto_bigint::{risc0, Encoding, U256};
+
+#[cfg(feature = "zeroize")]
+use zeroize::Zeroize;
+
+use crate::constants;
+
+/// Multiplicative Inverse of R mod L where R is the Montgomery modulus 2^261
+const R_INVERSE: U256 =
+    U256::from_be_hex("064EDB637937F48C1B0A73AA1C7FD1B5FD934BE6D1D6D67AC7421B8F04C727E2");
+
+/// 2^256 mod L
+const TWO_POW_TWO_FIFTY_SIX: U256 =
+    U256::from_be_hex("0FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEC6EF5BF4737DCF70D6EC31748D98951D");
+
+/// The `ScalarR0` struct represents an element in \\(\mathbb{Z} / \ell\mathbb{Z}\\)
+#[derive(Copy, Clone)]
+pub struct ScalarR0(pub U256);
+
+impl Debug for ScalarR0 {
+    fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
+        write!(f, "ScalarR0: {:?}", &self.0)
+    }
+}
+
+#[cfg(feature = "zeroize")]
+impl Zeroize for ScalarR0 {
+    fn zeroize(&mut self) {
+        self.0.zeroize();
+    }
+}
+
+impl ScalarR0 {
+    /// The scalar \\( -1 mod L \\).
+    pub const MINUS_ONE: ScalarR0 = ScalarR0(U256::from_be_hex(
+        "1000000000000000000000000000000014DEF9DEA2F79CD65812631A5CF5D3EC",
+    ));
+
+    /// Unpack a 32 byte / 256 bit scalar.
+    pub fn from_bytes(bytes: &[u8; 32]) -> ScalarR0 {
+        ScalarR0(U256::from_le_bytes(*bytes))
+    }
+
+    /// Reduce a 64 byte / 512 bit scalar mod l.
+    pub fn from_bytes_wide(bytes: &[u8; 64]) -> ScalarR0 {
+        let lo: U256 = U256::from_le_bytes(
+            bytes[0..32]
+                .try_into()
+                .expect("unable to parse low 32 bytes"),
+        );
+        let hi: U256 = U256::from_le_bytes(
+            bytes[32..]
+                .try_into()
+                .expect("unable to parse high 32 bytes"),
+        );
+
+        let hi_shifted_left_256 = risc0::modmul_u256(&hi, &TWO_POW_TWO_FIFTY_SIX, &constants::L.0);
+        // add_mod assumes the lhs + rhs is less than 2p. To guarantee this, we need to mod
+        // lo and hi by L
+        let lo = risc0::modmul_u256(&lo, &U256::ONE, &constants::L.0);
+        let total = hi_shifted_left_256.add_mod(&lo, &constants::L.0);
+
+        ScalarR0(total)
+    }
+
+    /// Pack the limbs of this `ScalarR0` into 32 bytes.
+    #[allow(clippy::identity_op)]
+    pub fn as_bytes(&self) -> [u8; 32] {
+        let val = risc0::modmul_u256(&self.0, &U256::ONE, &constants::L.0);
+        val.to_le_bytes()
+    }
+
+    /// Compute `a + b` (mod l).
+    pub fn add(a: &ScalarR0, b: &ScalarR0) -> ScalarR0 {
+        let result = a.0.add_mod(&b.0, &constants::L.0);
+        ScalarR0(result)
+    }
+
+    /// Compute `a - b` (mod l).
+    pub fn sub(a: &ScalarR0, b: &ScalarR0) -> ScalarR0 {
+        let result = a.0.sub_mod(&b.0, &constants::L.0);
+        ScalarR0(result)
+    }
+
+    /// Compute `-1 * a` (mod l).
+    pub fn negate(a: &ScalarR0) -> ScalarR0 {
+        let result = risc0::modmul_u256_denormalized(&a.0, &Self::MINUS_ONE.0, &constants::L.0);
```

Comment:
Other code in the `Scalar` implementation assumes the value is normalized, so you need to use the normalized version of this function. This comment applies to usages below as well.
```suggestion
        let result = risc0::modmul_u256(&a.0, &Self::MINUS_ONE.0, &constants::L.0);
```

Reply:
reverted all back to modmul_u256